File: 자료구조/나영/BOJ_10866_덱.java
import java.util.*;
import java.lang.*;
import java.io.*;

// 	16504 KB 	124 ms
class Main {
    static Deque<Integer> q = new LinkedList<>();
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static int n;
    static String s;
    static StringBuilder sb = new StringBuilder();
    public static void main(String[] args) throws IOException {
        n = Integer.parseInt(br.readLine());

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            s = st.nextToken();

            if (s.equals("push_front")) q.offerFirst(Integer.parseInt(st.nextToken()));
            else if (s.equals("push_back")) q.offerLast(Integer.parseInt(st.nextToken()));
            else if (s.equals("pop_front")) {
                if (!q.isEmpty()) sb.append(q.pollFirst()).append("\n");
                else sb.append(-1).append("\n");
            } else if (s.equals("pop_back")) {
                if (!q.isEmpty()) sb.append(q.pollLast()).append("\n");
                else sb.append(-1).append("\n");
            } else if (s.equals("size")) {
                sb.append(q.size()).append("\n");
            } else if (s.equals("empty")) {
                if (!q.isEmpty()) sb.append(0).append("\n");
                else sb.append(1).append("\n");
            } else if (s.equals("front")) {
                if (!q.isEmpty()) sb.append(q.peekFirst()).append("\n");
                else sb.append(-1).append("\n");
            } else {
                if (!q.isEmpty()) sb.append(q.peekLast()).append("\n");
                else sb.append(-1).append("\n");
            }
            
        }
        
        System.out.println(sb.toString());
    }
}

File: 자료구조/지우/BOJ_10866_덱.cpp
#include <iostream>
#include <deque>

using namespace std;

int main() {
    cin.tie(NULL);
    
    int N;
    cin >> N;
    
    // queue<int> q; // 와 큐는 push_back(), pop_back()이 안 먹네
    deque<int> dq; // 모든 입구에서 삽입 삭제 ㄱㄴ (큐+스택)
    
    for(int i=0; i<N; i++) {
        string s; int num;
        cin >> s;
        
        if(s=="push_back") {
            cin >> num;
            dq.push_back(num);
        } else if(s == "push_front") {
            cin >> num;
            dq.push_front(num);
        } else if(s == "pop_front") {
            if(dq.empty()) {
                cout << -1 << "\n";
            } else {
                cout << dq.front() << "\n"; 
                dq.pop_front();
            }
        } else if(s == "pop_back") {
            if(dq.empty()) {
                cout << -1 << "\n";
            } else {
                cout << dq.back() << "\n"; 
                dq.pop_back();
            }        
        } else if(s == "size") {
            cout << dq.size() << "\n";
        } else if(s == "empty") {
            if(dq.empty()) cout << 1 << "\n";
            else cout << 0 << "\n";
        } else if(s == "front") {
            if(dq.empty()) cout << -1 << "\n";
            else cout << dq.front() << "\n";
        } else if(s == "back") {
            if(dq.empty()) cout << -1 << "\n";
            else cout << dq.back() << "\n";
        } 
    }
    
}